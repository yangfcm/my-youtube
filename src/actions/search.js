import { defAxios as axios } from "../settings";
import { SEARCH, CATCH_ERROR } from "./types";
import { DEFAULT_ERROR_MSG } from "./default-error-msg";

/**
 * Search resources (including videos, channels, play lists)
 */
export const searchVideos = (filter, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/search", {
        params: {
          ...axios.defaults.params,
          ...filter,
          part: "snippet",
          maxResults: 15,
          pageToken,
        },
      });
      dispatch({
        type: SEARCH,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: {
          ...e.response.data.error,
          displayMessage: DEFAULT_ERROR_MSG.FAILED_TO_SERCH,
        },
      });
    }
  };
};
