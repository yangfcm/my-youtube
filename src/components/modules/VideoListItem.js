import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const VideoListItem = ({ video, playlistId }) => {
  // console.log(video);
  let linkUrl = video.contentDetails
    ? `/video/${video.contentDetails.videoId}`
    : `/video/${video.id.videoId}`;
  if (playlistId) linkUrl += `?playlistId=${playlistId}`;
  return (
    <div className="d-flex flex-md-row my-2">
      <div className="d-flex align-items-center">
        <Link to={linkUrl}>
          <img
            src={
              video.snippet.thumbnails && video.snippet.thumbnails.medium.url
            }
            alt={video.title}
            style={{ maxWidth: "140px" }}
          />
        </Link>
      </div>
      <div className="d-flex flex-column justify-content-center ml-3">
        <Link to={linkUrl} title={video.snippet.title}>
          <div
            className="text-primary font-weight-bold"
            style={{
              lineHeight: "23px",
              maxHeight: "46px",
              overflow: "hidden",
              textOverflow: "ellipsis",
            }}
          >
            {video.snippet.title}
          </div>
        </Link>
        <Link to={`/channel/${video.snippet.channelId}`}>
          {" "}
          {video.snippet.channelTitle}
        </Link>
        <p className="font-weight-light">
          <FontAwesomeIcon icon="user-clock" />{" "}
          {moment(video.snippet.publishedAt).format("D MMM YYYY k:mm")}
        </p>
      </div>
    </div>
  );
};

export default VideoListItem;
