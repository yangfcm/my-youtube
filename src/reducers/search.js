import { SEARCH } from "../actions/types";

export const searchReducer = (state = {}, action) => {
  switch (action.type) {
    case SEARCH:
      return {
        ...state,
        searchResults: action.payload,
      };
    default:
      return state;
  }
};
