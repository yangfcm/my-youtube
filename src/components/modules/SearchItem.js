import React from "react";
import { Link } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import VideoListItem from "./VideoListItem";

/** Search result has three types: video, playlist and channel
 * Render them using different layout and styles based on their types
 */
const renderVideoItem = (video) => {
  return <VideoListItem video={video} />;
};

const renderChannelItem = (channel) => {
  const linkUrl = `/channel/${channel.id.channelId}`;
  return (
    <div className="d-flex flex-md-row my-3">
      <div className=" d-flex align-items-center justify-content-center">
        <Link to={linkUrl}>
          <img
            src={
              channel.snippet.thumbnails &&
              channel.snippet.thumbnails.medium.url
            }
            alt={channel.title}
            style={{ borderRadius: "50%", maxWidth: "120px" }}
          />
        </Link>
      </div>
      <div className=" d-flex flex-column justify-content-center ml-3">
        <Link to={linkUrl}>
          <h6 className="text-primary font-weight-bold">
            {channel.snippet.title}
          </h6>
        </Link>
        <p className="font-weight-light">{channel.snippet.description}</p>
      </div>
    </div>
  );
};

const renderPlaylistItem = (playlist) => {
  return (
    <div className="d-flex flex-md-row my-2">
      <div
        className="d-flex align-items-center"
        style={{ position: "relative" }}
      >
        <Link to={`/playlist/${playlist.id.playlistId}`}>
          <img
            src={
              playlist.snippet.thumbnails &&
              playlist.snippet.thumbnails.medium.url
            }
            alt={playlist.title}
            style={{ maxWidth: "140px" }}
          />
          <div
            style={{
              position: "absolute",
              zIndex: "10",
              height: "100%",
              width: "100%",
              background: "rgba(0,0,0, .4)",
              color: "white",
              top: "0",
            }}
            className="d-flex justify-content-center align-items-center"
          >
            {" "}
            <FontAwesomeIcon icon={"list"} />
            {"  "} Play List
          </div>
        </Link>
      </div>
      <div className="d-flex flex-column justify-content-center ml-3">
        <Link to={`/playlist/${playlist.id.playlistId}`}>
          <h6 className="text-primary font-weight-bold">
            {playlist.snippet.title}
          </h6>
        </Link>
        <Link to={`/channel/${playlist.snippet.channelId}`}>
          {" "}
          {playlist.snippet.channelTitle}
        </Link>
      </div>
    </div>
  );
};

const SearchItem = ({ item }) => {
  // console.log(item);
  const kind = item.id.kind.split("#")[1];
  return (
    <React.Fragment>
      {kind === "video" && renderVideoItem(item)}
      {kind === "channel" && renderChannelItem(item)}
      {kind === "playlist" && renderPlaylistItem(item)}
    </React.Fragment>
  );
};

export default SearchItem;
