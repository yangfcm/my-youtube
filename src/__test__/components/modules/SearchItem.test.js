import React from "react";
import { shallow } from "enzyme";
import { Link } from "react-router-dom";
import SearchItem from "components/modules/SearchItem";
import VideoListItem from "components/modules/VideoListItem";
import { searchResultResponse } from "__test__/fixtures/search";

describe("Test SearchItem component", () => {
  let wrapper;
  const videoSearchResult = searchResultResponse.items.filter(
    (item) => item.id.kind === "youtube#video"
  )[0];
  const playlistSearchResult = searchResultResponse.items.filter(
    (item) => item.id.kind === "youtube#playlist"
  )[0];
  const channelSearchResult = searchResultResponse.items.filter(
    (item) => item.id.kind === "youtube#channel"
  )[0];

  describe("When searched item is video", () => {
    beforeEach(() => {
      wrapper = shallow(<SearchItem item={videoSearchResult} />);
    });

    afterEach(() => {
      wrapper.unmount();
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render VideoListItem component", () => {
      expect(wrapper.find(VideoListItem).exists()).toBe(true);
      expect(wrapper.find(VideoListItem).prop("video")).toEqual(
        videoSearchResult
      );
    });
  });

  describe("When serachedItem is playlist", () => {
    beforeEach(() => {
      wrapper = shallow(<SearchItem item={playlistSearchResult} />);
    });

    afterEach(() => {
      wrapper.unmount();
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render correct image", () => {
      expect(wrapper.find("img").prop("src")).toBe(
        playlistSearchResult.snippet.thumbnails.medium.url
      );
    });

    it("should render correct link", () => {
      const links = wrapper.find(Link);
      expect(links.at(0).prop("to")).toBe(
        `/playlist/${playlistSearchResult.id.playlistId}`
      );
      expect(links.at(1).prop("to")).toBe(
        `/playlist/${playlistSearchResult.id.playlistId}`
      );
      expect(links.at(2).prop("to")).toBe(
        `/channel/${playlistSearchResult.snippet.channelId}`
      );
      expect(links.at(2).text()).toContain(
        playlistSearchResult.snippet.channelTitle
      );
    });

    it("should render correct playlist title", () => {
      expect(wrapper.find("h6").text()).toBe(
        playlistSearchResult.snippet.title
      );
    });
  });

  describe("When searchedItem is channel", () => {
    beforeEach(() => {
      wrapper = shallow(<SearchItem item={channelSearchResult} />);
    });

    afterEach(() => {
      wrapper.unmount();
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render correct image", () => {
      expect(wrapper.find("img").prop("src")).toBe(
        channelSearchResult.snippet.thumbnails.medium.url
      );
    });

    it("should render correct link", () => {
      const links = wrapper.find(Link);
      expect(links.at(0).prop("to")).toBe(
        `/channel/${channelSearchResult.id.channelId}`
      );
      expect(links.at(1).prop("to")).toBe(
        `/channel/${channelSearchResult.id.channelId}`
      );
    });

    it("should render channel title", () => {
      expect(wrapper.find("h6").text()).toContain(
        channelSearchResult.snippet.title
      );
    });

    it("should render title description", () => {
      expect(wrapper.find("p.font-weight-light").text()).toContain(
        channelSearchResult.snippet.description
      );
    });
  });
});
