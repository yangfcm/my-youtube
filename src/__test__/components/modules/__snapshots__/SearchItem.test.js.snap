// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SearchItem component When searched item is video should render component 1`] = `
<Fragment>
  <VideoListItem
    video={
      Object {
        "etag": "fUy83ga9B0qPSeFNWMj7RhWuLUY",
        "id": Object {
          "kind": "youtube#video",
          "videoId": "H89hKw06iWs",
        },
        "kind": "youtube#searchResult",
        "snippet": Object {
          "channelId": "UCnUYZLuoy1rq1aVMwx4aTzw",
          "channelTitle": "Google Chrome Developers",
          "description": "Join us for day 1 of web.dev LIVE as we head to the Americas timezone to provide helpful content on Performance and Discovery through Search. Enjoy the ...",
          "liveBroadcastContent": "none",
          "publishTime": "2020-06-30T16:00:10Z",
          "publishedAt": "2020-06-30T16:00:10Z",
          "thumbnails": Object {
            "default": Object {
              "height": 90,
              "url": "https://i.ytimg.com/vi/H89hKw06iWs/default.jpg",
              "width": 120,
            },
            "high": Object {
              "height": 360,
              "url": "https://i.ytimg.com/vi/H89hKw06iWs/hqdefault.jpg",
              "width": 480,
            },
            "medium": Object {
              "height": 180,
              "url": "https://i.ytimg.com/vi/H89hKw06iWs/mqdefault.jpg",
              "width": 320,
            },
          },
          "title": "web.dev LIVE 2020: Day One",
        },
      }
    }
  />
</Fragment>
`;

exports[`Test SearchItem component When searchedItem is channel should render component 1`] = `
<Fragment>
  <div
    className="d-flex flex-md-row my-3"
  >
    <div
      className=" d-flex align-items-center justify-content-center"
    >
      <Link
        to="/channel/UC_x5XG1OV2P6uZZ5FSM9Ttw"
      >
        <img
          src="https://yt3.ggpht.com/-Fgp8KFpgQqE/AAAAAAAAAAI/AAAAAAAAAAA/Wyh1vV5Up0I/s240-c-k-no-mo-rj-c0xffffff/photo.jpg"
          style={
            Object {
              "borderRadius": "50%",
              "maxWidth": "120px",
            }
          }
        />
      </Link>
    </div>
    <div
      className=" d-flex flex-column justify-content-center ml-3"
    >
      <Link
        to="/channel/UC_x5XG1OV2P6uZZ5FSM9Ttw"
      >
        <h6
          className="text-primary font-weight-bold"
        >
          Google Developers
        </h6>
      </Link>
      <p
        className="font-weight-light"
      >
        The Google Developers channel features talks from events, educational series, best practices, tips, and the latest updates across our products and platforms.
      </p>
    </div>
  </div>
</Fragment>
`;

exports[`Test SearchItem component When serachedItem is playlist should render component 1`] = `
<Fragment>
  <div
    className="d-flex flex-md-row my-2"
  >
    <div
      className="d-flex align-items-center"
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <Link
        to="/playlist/IDZyZMYn3H0"
      >
        <img
          src="https://i.ytimg.com/vi/IDZyZMYn3H0/mqdefault.jpg"
          style={
            Object {
              "maxWidth": "140px",
            }
          }
        />
        <div
          className="d-flex justify-content-center align-items-center"
          style={
            Object {
              "background": "rgba(0,0,0, .4)",
              "color": "white",
              "height": "100%",
              "position": "absolute",
              "top": "0",
              "width": "100%",
              "zIndex": "10",
            }
          }
        >
           
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="list"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
            
           Play List
        </div>
      </Link>
    </div>
    <div
      className="d-flex flex-column justify-content-center ml-3"
    >
      <Link
        to="/playlist/IDZyZMYn3H0"
      >
        <h6
          className="text-primary font-weight-bold"
        >
          How to Become a Google Developer Expert (GDE)
        </h6>
      </Link>
      <Link
        to="/channel/UCsBjURrPoezykLs9EqgamOA"
      >
         
        Fireship
      </Link>
    </div>
  </div>
</Fragment>
`;
